tensorflow/python/BUILD

```diff
@@ -5976,11 +5976,10 @@ py_library(
     name = "pywrap_tensorflow",
     srcs = [
         "pywrap_tensorflow.py",
-    ] + if_static(
-        ["pywrap_dlopen_global_flags.py"],
-        # Import will fail, indicating no global dlopen flags
-        otherwise = [],
-    ),  # b/153585257
+        # modular TF need this file to load expose C API symbols
+        # in pywrap_tensorflow_internal.so
+        "pywrap_dlopen_global_flags.py",
+    ],
```

Comment:
Instead, I think we should make the right symbols available in libtensorflow_framework.so. For example, you can take a look how filesystem plugin symbols are added:
1. Add headers to pip package: https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/tools/pip_package/BUILD;l=39?q=%22tensorflow%2Fc%2Fexperimental%2Ffilesystem:filesystem_interface%22
2. Add headers to libtensorflow_framework:
https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/BUILD;l=708
3. I think the dependency on :modular_filesystem here adds implementation to pywrap_tensorflow.
https://github.com/tensorflow/tensorflow/pull/43680/files

Let me know and I can help out add these dependencies or you can add them in this change.

Reply:
we find it's a little complex to make it similar as modular_filesystem, because it not only contains stream executor c api , but also those kernel c api and op c api. If you can help on these dependencies, that will be appreciate.

Reply:
Sounds good, I will submit a change to expose kernel and op C APIs.

Reply:
I exposed all 3 dependencies (stream_executor.h, ops.h, kernels.h) here: https://github.com/tensorflow/tensorflow/commit/c0904b922dd6e4c30f94932252e3fe7a8ecfa43e

Reply:
@jzhoulon can you check if you can remove this change now? I updated stream_executor_hdrs to only depend on headers now:
https://github.com/tensorflow/tensorflow/commit/47b9d665929b8a313ca56dcee6c796155a1851a8

Reply:
Just a quick update that I can repro the issue and looking into ways we can fix it.
From what I heard, we don't use pywrap_dlopen_global_flags.py with dynamic linking because we used to get symbol conflicts.

Reply:
thanks for the update, if there is any thing we need to change, please let us know. thanks.

Reply:
I just tried following instructions for custom op and it seems to be working at least for kernel API. Although, I still think this won't work for stream executor APIs as I still need to `SP_*` and `SE_*` to version script here: https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/tf_version_script.lds.

Basically, I followed https://www.tensorflow.org/guide/create_op for a dummy op implementation:
test.cc
```cpp
#include <stdio.h>
#include <tensorflow/c/kernels.h>
#include <tensorflow/c/tf_status.h>

static void* FooCreate(TF_OpKernelConstruction* ctx) { return nullptr; }
static void FooDelete(void* kernel) {}
static void FooCompute(void* kernel, TF_OpKernelContext* ctx) {}

void RegisterOpKernel() {
  TF_Status* status = TF_NewStatus();
  {
    auto* builder = TF_NewKernelBuilder("Bitcast", "CPU", &FooCreate,
                                        &FooCompute, &FooDelete);
    TF_RegisterKernelBuilder("BitcastOp", builder, status);
  }
  TF_DeleteStatus(status);
}

__attribute__((unused)) static bool IsOpKernelRegistered = []() {
  RegisterOpKernel();
  return true;
}();
```
I install TF with:
```
python3 -m venv venv
source venv/bin/activate
pip install tf-nightly
```

Then, build it with:
```
g++ -std=c++11 -shared test.cc -o test.so -fPIC -Ivenv/lib/python3.8/site-packages/tensorflow/include -Lvenv/lib/python3.8/site-packages/tensorflow/python -l:_pywrap_tensorflow_internal.so -O2
```
Note that I am passing `_pywrap_tensorflow_internal.so` here unlike in custom op documentation.

Then, run:
```
cp test.so venv/lib/python3.8/site-packages/tensorflow-plugins/
```
Now, `import tensorflow as tf` works just fine without plugin loading errors.
I will add stream executor symbols to version script soon, but in meantime let me know if I am missing something with this example.

Reply:
@annarev, I saw there is a explicit dependency "-L pywrap_tensorflow_internal.so", does it mean when we build plugin, we need always build tensorflow first, then copy or add the _pywrap_tensorflow_internal.so to the plugin's build dependency path? In my opinion , this may not be a good use case for plugin developers since they belong to different project repo. 

if we can have a way to support loading pywrap_tensorflow_internal.so with RTLD_GLOBAL  or others ways, so plugin build will not depend on tensorflow, I think this may make the things simpler. Thanks.
```
g++ -std=c++11 -shared test.cc -o test.so -fPIC -Ivenv/lib/python3.8/site-packages/tensorflow/include -Lvenv/lib/python3.8/site-packages/tensorflow/python -l:_pywrap_tensorflow_internal.so -O2
```

Reply:
I was thinking you can just `pip install tf-nightly` and link against `site-packages/tensorflow/python/_pywrap_tensorflow_internal.so` when I wrote the instructions.
If this is not preferable, can you help me understand better the challenges of using this approach?

Reply:
ok, thanks for the explain.  I will remove this modification(pywrap_dlopen_global_flags.py).

Reply:
@annarev I have reverted this change. If you finished the "SP_* and SE_* to version script", please let me know and I can try if there is any other issue. Thanks.

Reply:
`SP_` and `SE_` symbols have been added: https://github.com/tensorflow/tensorflow/commit/f296fd6f747196e2a527b53b3f62cec078048bbd. Thank you for checking!

Reply:
@annarev We try this and it seems worked, thanks for the help.

Reply:
Good to hear, thank you for confirming!
Just a heads up, I am still looking at best way to do it. We will probably switch to linking to libtensorflow_framework.so instead of _pywrap_tensorflow_internal.so. This would let us use the plugin with other languages, not just Python.